src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,1366 @@
+	if (HeapTupleIsValid(tp))
+	{
+		Form_pg_type typtup = (Form_pg_type) GETSTRUCT(tp);
+
+		result = typtup->typowner;
+		ReleaseSysCache(tp);
+	}
+
+	return result;
+}
+
+
+/*
+ * makeTypeNameFromRangeVar creates a TypeName based on a RangeVar.
+ */
+static TypeName *
+makeTypeNameFromRangeVar(const RangeVar *relation)
+{
+	List *names = NIL;
+	if (relation->schemaname)
+	{
+		names = lappend(names, makeString(relation->schemaname));
+	}
+	names = lappend(names, makeString(relation->relname));
+
+	return makeTypeNameFromNameList(names);
+}
+
+
+/*
+ * EnsureSequentialModeForTypeDDL makes sure that the current transaction is already in
+ * sequential mode, or can still safely be put in sequential mode, it errors if that is
+ * not possible. The error contains information for the user to retry the transaction with
+ * sequential mode set from the beginnig.
+ *
+ * As types are node scoped objects there exists only 1 instance of the type used by
+ * potentially multiple shards. To make sure all shards in the transaction can interact
+ * with the type the type needs to be visible on all connections used by the transaction,
+ * meaning we can only use 1 connection per node.
+ */
+static void
+EnsureSequentialModeForTypeDDL(void)
+{
+	if (!IsTransactionBlock())
+	{
+		/* we do not need to switch to sequential mode if we are not in a transaction */
+		return;
+	}
+
+	if (ParallelQueryExecutedInTransaction())
+	{
+		ereport(ERROR, (errmsg("cannot create or modify type because there was a "
+							   "parallel operation on a distributed table in the "
+							   "transaction"),
+						errdetail("When creating or altering a type, Citus needs to "
+								  "perform all operations over a single connection per "
+								  "node to ensure consistency."),
+						errhint("Try re-running the transaction with "
+								"\"SET LOCAL citus.multi_shard_modify_mode TO "
+								"\'sequential\';\"")));
+	}
+
+	ereport(DEBUG1, (errmsg("switching to sequential query execution mode"),
+					 errdetail("Type is created or altered. To make sure subsequent "
+							   "commands see the type correctly we need to make sure to "
+							   "use only one connection for all future commands")));
+	SetLocalMultiShardModifyModeToSequential();
+}
+
+
+/*
+ * ShouldPropagateTypeCreate returns if we should propagate the creation of a type.
+ *
+ * There are two moments we decide to not directly propagate the creation of a type.
+ *  - During the creation of an Extension; we assume the type will be created by creating
+ *    the extension on the worker
+ *  - During a transaction block; if types are used in a distributed table in the same
+ *    block we can only provide parallelism on the table if we do not change to sequentia
+ *    mode. Types will be propagated outside of this transaction to the workers so that
+ *    the transaction can use 1 connection per shard and fully utilize citus' parallelism
+ */
+static bool
+ShouldPropagateTypeCreate()
+{
+	if (creating_extension)
```

Comment:
I think it could be valuable, but I guess we want this guc to guard more then the creation of a type.

It should also prevent the dependency traversal to types on distributing a table, adding a node etc. as well as alter type statements.

My understanding was that this guc should be able to NOOP this whole change

Reply:
> My understanding was that this guc should be able to NOOP this whole change

Yes, and also we can do it after merging this, but before 9.0 with a separate PR?

Reply:
makes sense

Reply:
issue for tracking: https://github.com/citusdata/citus/issues/2954